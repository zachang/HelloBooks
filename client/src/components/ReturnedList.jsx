import React from 'react';
import PropTypes from 'react-proptypes';
import swal from 'sweetalert2';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Pagination } from 'react-materialize';
import { viewAllReturnedAction, confirmReturnAction } from '../actions/bookAction';
import AdminHeader from './common/AdminHeader.jsx';
import AdminSidebar from './common/AdminSidebar.jsx';
import UserReturnedList from './returns/UserReturnedList.jsx';

/**
 * ReturnedList class declaration
 * @class ReturnedList
 * @extends {React.Component}
 */
export class ReturnedList extends React.Component {
  /**
   * class constructor
   * @param {object} props
   */
  constructor(props) {
    super(props);
    this.state = {
      errors: null,
      showToast: false,
      pageCount: null,
      limit: 15,
      clickedReturnedBorrowList: []
    };
    this.confirmReturn = this.confirmReturn.bind(this);
  }

  /**
   * @method componentWillMount
   * @return {void} void
   */
  componentWillMount() {
    this.props.viewAllReturnedAction(this.state.limit, 0);
  }

  /**
   * @method componentWillReceiveProps
   * @param {object} nextProps - nextProps
   * @return {object} nextProps
   */
  componentWillReceiveProps(nextProps) {
    this.setState({ clickedReturnedBorrowList: nextProps.bookState.clickedReturnedBorrowList });
    if (nextProps.bookState.success === false) {
      this.setState({ errors: nextProps.bookState.errors });
    }
    if (this.state.pageCount !== nextProps.bookState.pageCount) {
      this.setState({ pageCount: nextProps.bookState.pageCount });
    }
  }

  /**
   * Handles  book return confirmation
   * @method confirmReturn
   * @return {void}
   * @param {integer} borrowId - borrowId
   */
  confirmReturn(borrowId) {
    swal({
      title: 'Are you sure?',
      text: "You won't be able to revert this!",
      type: 'warning',
      showCancelButton: true,
      confirmButtonColor: '#26a69a',
      cancelButtonColor: '#d33',
      confirmButtonText: 'Confirm!'
    }).then((result) => {
      if (result.value) {
        swal(
          'Confirmed!',
          'Return has been confirmed.',
          'success'
        );
        this.props.confirmReturnAction(borrowId);
      }
    });
    this.setState({ showToast: true });
  }

  /**
   * Renders AddBook component
   * @return {XML} JSX
   */
  render() {
    return (
      <div className='row'>
        <AdminHeader/>
        <AdminSidebar/>
        <div className='container mainCon' style={{ marginLeft: '5%' }}>
          <div className='row'>
            <div className='section'>
              <h4 style={{ marginTop: '7%' }}>Returned Books</h4>
            </div>
            <div className='divider' style={{ width: '100%', marginTop: '-2.5%', marginBottom: '3.5%' }}></div>

            <div className='row'>
              <table className='bordered highlight'>
                <thead className='black white-text'>
                  <tr>
                    <th>Book Name</th>
                    <th>Borrower</th>
                    <th>Borrow_Date</th>
                    <th>Expected_Return_Date</th>
                    <th>Return_Date</th>
                    <th>Overdue</th>
                    <th>Confirm</th>
                  </tr>
                </thead>

                <tbody>
                  { this.props.bookState.returners.map((returner, i) =>
                    <UserReturnedList
                      key={i}
                      returner={returner}
                      clickedReturnedBorrowList={this.state.clickedReturnedBorrowList}
                      confirmReturn={this.confirmReturn}
                    />
                  )}
                </tbody>
              </table>
            </div>

            <div className='row'>
              {
                ((this.state.pageCount) ?
                  <Pagination
                    items={this.state.pageCount}
                    onSelect={(page) => {
                      const offset = (page - 1) * this.state.limit;
                      this.props.viewAllReturnedAction(this.state.limit, offset);
                    }} /> : '')
              }
            </div>
          </div>
        </div>
      </div>

    );
  }
}

ReturnedList.propTypes = {
  bookState: PropTypes.object.isRequired,
  viewAllReturnedAction: PropTypes.func.isRequired,
  confirmReturnAction: PropTypes.func.isRequired
};
const mapStateToProps = state => ({
  bookState: state.bookReducer,
});
const mapDispatchToProps = dispatch =>
  bindActionCreators({ viewAllReturnedAction, confirmReturnAction }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(ReturnedList);