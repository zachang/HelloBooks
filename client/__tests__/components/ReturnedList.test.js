import React from 'react';
import toJson from 'enzyme-to-json';
import sinon from 'sinon';
import { shallow } from 'enzyme';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import storeConfiguration from '../../src/store/store';
import { ReturnedList } from '../../src/components/ReturnedList.jsx';

describe('<ReturnedList/>', () => {
  // create a spy function for viewAllReturnedAction
  const viewAllReturnedAction = sinon.spy();
  // create a spy function for confirmReturnAction
  const confirmReturnAction = sinon.spy();
  // spy on confirmReturn of ReturnedList Page
  sinon.spy(ReturnedList.prototype, 'confirmReturn');
  // spy on componentWillMount of ReturnedList Page
  sinon.spy(ReturnedList.prototype, 'componentWillMount');
  // spy on componentWillReceiveProps of ReturnedList Page
  sinon.spy(ReturnedList.prototype, 'componentWillReceiveProps');

  const props = {
    bookState: {
      success: false,
      passes: null,
      errors: null,
      fails: null,
      books: [],
      book: null,
      borrows: null,
      allBorrows: null,
      borrowers: [],
      returned: null,
      returnings: null,
      returners: ['mmmm'],
      pageCount: null,
      confirmReturn: null,
      clickedReturnedBorrowList: [],
      confirmBorrow: null,
      clickedBorrowList: []
    },
    viewAllReturnedAction,
    confirmReturnAction
  };

  const middleware = [thunk];
  const mockStore = configureMockStore(middleware);
  let store;
  let shallowComponent;
  beforeEach(() => {
    store = mockStore(storeConfiguration);
    shallowComponent = shallow(<ReturnedList {...props} />);
    shallowComponent.setState({
      pageCount: 1,
      returners: ['mmm', 'mmmm'],
      showToast: false,
      clickedReturnedBorrowList: []
    });
  });

  it('renders <ReturnedList /> component', () => {
    const tree = toJson(shallowComponent);
    expect(tree).toMatchSnapshot();
  });

  it('calls componentWillReceiveProps when paginating', () => {
    props.pageCount = 1;
    props.bookState.pageCount = 2;
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(ReturnedList.prototype.componentWillReceiveProps.called)
      .toEqual(true);
  });

  it(`should check if the 'Pagination' is defined`, () => {
    expect(shallowComponent.find('Pagination').length).toBe(1);
    shallowComponent.find('Pagination').simulate('select');
  });

  it('calls componentWillReceiveProps when paginating', () => {
    props.bookState.sucess = false;
    shallowComponent.setState({
      errors: props.bookState.errors
    });
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(ReturnedList.prototype.componentWillReceiveProps.called)
      .toEqual(true);
  });

  it('should check that confirmReturn is called', () => {
    shallowComponent.instance().confirmReturn(1);
    expect(ReturnedList.prototype.confirmReturn.called).toEqual(true);
  });
});