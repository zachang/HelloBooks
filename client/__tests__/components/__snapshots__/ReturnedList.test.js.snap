// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReturnedList/> renders <ReturnedList /> component 1`] = `
<div
  className="row"
>
  <div
    className="section"
  >
    <h4
      style={
        Object {
          "marginTop": "7%",
        }
      }
    >
      Returned Books
    </h4>
  </div>
  <div
    className="divider"
    style={
      Object {
        "marginBottom": "3.5%",
        "marginTop": "-2.5%",
        "width": "100%",
      }
    }
  />
  <div
    className="row"
  >
    <table
      className="bordered highlight"
    >
      <thead
        className="black white-text"
      >
        <tr>
          <th>
            Book Name
          </th>
          <th>
            Borrower
          </th>
          <th>
            Borrow_Date
          </th>
          <th>
            Expected_Return_Date
          </th>
          <th>
            Return_Date
          </th>
          <th>
            Overdue
          </th>
          <th>
            Confirm
          </th>
        </tr>
      </thead>
      <tbody>
        <UserReturnedList
          clickedReturnedBorrowList={Array []}
          confirmReturn={[Function]}
          key="0"
          returner="mmmm"
        />
      </tbody>
    </table>
  </div>
  <div
    className="row"
  >
    <Pagination
      activePage={1}
      items={1}
      onSelect={[Function]}
    />
  </div>
</div>
`;
